### PR TITLE
Undo breaking API change in unitary synthesis plugin interface

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit reverts an unecessary breaking API change that was introduced in #9175. In that PR the data type for two arguments, gate_lengths and gate_errors, that was potentially passed to plugins was changed. This change was well intentioned as it was used by the default plugin as part of the changes to the default plugin made in that PR. But this neglected the downstream effects for any existing plugin interface users relying on the input to those fields. Making this change to existing fields would break any plugins that were using the old data format and is a violation of the Qiskit stability and deprecation policy. This commit reverts that change so that gate_lengths and gate_errors contain the same data format as in previous releases. Instead to facilitate the new workflow a new fields, gate_lengths_by_qubit and gate_errors_by_qubit, were added that leverage the new format introduced by #9175. By adding this as new optional fields existing users can continue to work as before, but the default plugin which requires the different data format can use the new data type.

### CODE REVIEW

**Comment by ajavadia** [comment]
I did add the API change label and upgrade releasenote in that PR.
Are there synthesis plugins that actually use these gate errors/lengths? I'm afraid this migration path from uniform gates to heterogeous gates will be very slow and protracted if we keep adding code without removing old ones.

**Comment by mtreinish** [comment]
Yeah, you did the right things to document a breaking change, but that doesn't mean we really should be breaking the API in this case. I don't really know if there are any plugins out there using these interfaces either, but to a certain degree we can't really know for sure because there is no requirement for any downstream plugin interface users to publish or document their usage. If there isn't any value in having the old view moving forward we can deprecate it in 0.25.0 (we have to release the alternative first before marking it as deprecated). The key though is that we give any plugin authors/maintainers a chance to migrate before removing the API they've potentially been relying on.

 Although, in this case I'm personally unclear if there is any harm in keeping both views around long term there really isn't a guarantee that a plugin supports synthesizing to heterogeneous basis gates (lol or that it even respects the basis gates set in the target), but we can discuss that in a deprecation PR.

File: qiskit/transpiler/passes/synthesis/unitary_synthesis.py [hunk 1]
```diff
@@ -479,6 +490,46 @@ def _recurse(dag):
 
 
 def _build_gate_lengths(props=None, target=None):
+    gate_lengths = {}
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/transpiler/passes/synthesis/unitary_synthesis.py:502]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10066/commits/a2bb6e196141b00443653af41090d182106c0c7a

File: qiskit/transpiler/passes/synthesis/plugin.py [hunk 2]
```diff
@@ -341,8 +349,7 @@ def supports_pulse_optimize(self):
         pass
 
     @property
-    @abc.abstractmethod
-    def supports_gate_lengths(self):
+    def supports_gate_lengths_by_qubit(self):
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/transpiler/passes/synthesis/plugin.py:352]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10066/commits/a2bb6e196141b00443653af41090d182106c0c7a

**Review by kdk — APPROVED** [review]

**Pull request merged**